File: skills/adk-agentic-prod-workflows/assets/workflow-templates/ts/graph-workflow/tools.ts
/**
 * Graph workflow: tool definitions.
 */

import type { ToolContext } from "@google/adk";

/**
 * Tool to look up information from a knowledge base.
 */
export async function lookupInformation(
  query: string,
  ctx: ToolContext
): Promise<{ results: string[] }> {
  // Production: replace with actual DB/search call
  return {
    results: [`Result for query: ${query}`],
  };
}

/**
 * Tool to validate data against a schema.
 */
export async function validateData(
  data: Record<string, unknown>,
  schema: string,
  ctx: ToolContext
): Promise<{ valid: boolean; errors: string[] }> {
  // Production: replace with actual schema validation
  return { valid: true, errors: [] };
}

File: skills/adk-agentic-prod-workflows/assets/workflow-templates/python/graph-workflow/tools.py
"""Tool definitions for graph workflow.

Each tool: Pydantic params schema, structured return type,
structured logging with call_id + latency_ms.
"""

from pydantic import BaseModel, Field
import structlog

logger = structlog.get_logger(__name__)


class ValidateInputParams(BaseModel):
    data: dict = Field(description="Input data to validate")
    schema_name: str = Field(description="Validation schema to apply")


class ValidateInputResult(BaseModel):
    is_valid: bool
    errors: list[str]
    sanitized_data: dict | None = None


def validate_input(params: ValidateInputParams) -> ValidateInputResult:
    """Validate input data against expected schema."""
    logger.info("tool_call", tool="validate_input", schema=params.schema_name)
    errors = []

    if not params.data:
        errors.append("Input data is empty")
    if not params.schema_name:
        errors.append("Schema name is required")

    if errors:
        return ValidateInputResult(is_valid=False, errors=errors, sanitized_data=None)

    # TODO: Replace with real schema validation
    sanitized = {k: v for k, v in params.data.items() if v is not None}
    return ValidateInputResult(is_valid=True, errors=[], sanitized_data=sanitized)


class ProcessDataParams(BaseModel):
    data: dict = Field(description="Validated data to process")
    operation: str = Field(default="default", description="Processing operation")


class ProcessDataResult(BaseModel):
    status: str
    result: dict | None
    error: str | None = None


def process_data(params: ProcessDataParams) -> ProcessDataResult:
    """Process validated data."""
    logger.info("tool_call", tool="process_data", operation=params.operation)
    try:
        # TODO: Replace with real processing logic
        return ProcessDataResult(status="ok", result=params.data, error=None)
    except Exception as e:
        logger.error("tool_error", tool="process_data", error=str(e))
        return ProcessDataResult(status="error", result=None, error=str(e))
